### PR TITLE
Mui admin app auth

### PR DESCRIPTION
**Pull request by sidravi1**
Reviewer: @amiraliemami 
Estimate: 2 hrs

---

## Ticket

Fixes: [408](https://idinsight.atlassian.net/browse/AAQ-408?atlOrigin=eyJpIjoiMmI2ZDQ5YjhjNzdjNDRjYzgwZWJmYTIxMTFiNzAyMTgiLCJwIjoiaiJ9)

### Goal

Setup auth/access controls for new MUI Admin app

### Changes

new AuthProvider component
new login screen
components disabled if "readonly"

## How has this been tested?

Only through the UI.

## Checklist

Fill with `x` for completed. Delete any lines that are not relevant

- [x] My code follows the style guidelines of this project
- [x] I have reviewed my own code to ensure good quality
- [x] I have tested the functionality of my code to ensure it works as intended
- [x] I have resolved merge conflicts
- [ ] I have updated the automated tests (if applicable)
- [ ] I have updated the requirements (if applicable)
- [ ] I have updated the README file (if applicable)
- [ ] I have updated affected documentation (if applicable)

### CODE REVIEW

File: admin_app_2/src/app/content/edit/page.tsx
Comment (amiraliemami) [inline admin_app_2/src/app/content/edit/page.tsx:34]: This entire page shouldn't run any code at all (including the .getContent fetch, since it's never shown) if we're readonly and instead display a "Not Authorised" message on the page.

How do we also stop the useEffects from running on page load based on auth state?

File: admin_app_2/src/app/content/edit/page.tsx
Comment (amiraliemami) [inline admin_app_2/src/app/content/edit/page.tsx:64]: This is cool!

File: admin_app_2/src/components/ProtectedComponent.tsx
Comment (amiraliemami) [inline admin_app_2/src/components/ProtectedComponent.tsx:34]: Display a message letting the user know they're off-limits.

```suggestion
    return <Layout.FlexBox
      sx={{
        justifyContent: 'center',
        alignItems: 'center',
        height: '70vh',
        typography: 'body1',
      }}
    >
      Not Authorised
    </Layout.FlexBox>;
```

File: admin_app_2/src/app/login/page.tsx
Comment (amiraliemami) [reply admin_app_2/src/app/login/page.tsx:1]: Works! Can you add a backlog ticket for making it look nicer?

File: admin_app_2/src/app/content/layout.tsx
Comment (amiraliemami) [inline admin_app_2/src/app/content/layout.tsx:3]: not used

File: admin_app_2/src/utils/auth.tsx
Comment (amiraliemami) [inline admin_app_2/src/utils/auth.tsx:35]: IMPORTANT: This default value resets the permission level on page reloads somehow. If I login with fullaccess and refresh a page it goes to readonly mode. If I change this default value to "fullaccess" and login as readonly, reloads reset permissions to fullaccess!

**Review by amiraliemami — CHANGES_REQUESTED** [review]
Thanks for this - learned a lot!

Left some comments - three major things I came across to fix before merge:
1. Refreshing the page resets the permissions to the default value set in the `auth.tsx` file (see comment on there).
2. Edit page's useEffect still runs and fetches content in the background even with just readonly access (see comment).
3. Edit and New buttons' links still work even on readonly. Suggested fixes.

Thanks!

File: admin_app_2/src/components/ContentModal.tsx
Comment (sidravi1) [reply admin_app_2/src/components/ContentModal.tsx:182]: I see it as more modularised! There are few important principles of good design at play:
* lower level components should be like pure functions. Fully deterministic by arguments. It makes it easier to test and reuse code.
* all dependencies should be at the same level of abstraction. We had some calls to api.tsx from `page.tsx` and some from `/components`. This makes it hard to reason changes.
* Coupling should be at a high level. All low level components should just use code within the module and should not know the inner workings of other module. Allows for easier refactors. 

Happy to talk more. There is a good chapter Pragmatic Programmer and Clean Code that I can try to dig up.

In the interest of time, I'd rather leave it here. You'll be reviewing it anyway and it's all going to get squashed as one change to main anyway in the end.

File: admin_app_2/src/app/login/page.tsx
Comment (sidravi1) [reply admin_app_2/src/app/login/page.tsx:1]: Done! [AAQ-419](https://idinsight.atlassian.net/browse/AAQ-419)

File: admin_app_2/src/components/ProtectedComponent.tsx
Comment (sidravi1) [reply admin_app_2/src/components/ProtectedComponent.tsx:31]: Done. [AAQ-420](https://idinsight.atlassian.net/browse/AAQ-420)

File: admin_app_2/src/utils/auth.tsx
Comment (sidravi1) [reply admin_app_2/src/utils/auth.tsx:53]: i thought `fromPage` made more sense? It's the page from where you came. But happy to change it to `sourcePage`

File: admin_app_2/src/utils/auth.tsx
Comment (sidravi1) [reply admin_app_2/src/utils/auth.tsx:30]: `useEffect` runs at each render of the component but since it's async, it is not guaranteed to be done when this is being initialized. So when you use `useAuth` it returns `null` since it hasn't updated when the other page (like /content) is being rendered.

I think we leave it as is for now.

File: admin_app_2/src/utils/auth.tsx
Comment (sidravi1) [reply admin_app_2/src/utils/auth.tsx:35]: Yes - because Auth context is lost on refresh!
I had a few options to fix this:
* log out on refresh - annoying
* save `accessLevel` to localstorage and load from from there. Not safe because user can just edit their local storage to go from `readonly` to `fullaccess`
* save `accessLevel` as an encrypted object with rotating keys. Oof. That's a big change.

I'm open to option 1 or 2. And I think I'll reach out to Griffin or Jay to get ideas on fixing this.

Raised a [ticket](https://idinsight.atlassian.net/browse/AAQ-421) for it.

File: admin_app_2/src/app/content/page.tsx
Comment (amiraliemami) [reply admin_app_2/src/app/content/page.tsx:57]: Yeah the button link is still active because of how the link is a separate component. Combining it as suggested disables it correctly.

File: admin_app_2/src/components/ContentModal.tsx
Comment (amiraliemami) [reply admin_app_2/src/components/ContentModal.tsx:182]: Makes sense, thanks for the explanation. That would've been easy for me to see on Python but it felt different here because it felt more unfamiliar. Thanks for picking this up!

File: admin_app_2/src/utils/auth.tsx
Comment (amiraliemami) [reply admin_app_2/src/utils/auth.tsx:53]: yeah I think source makes more sense since it's a noun, so that is the "sourcePage" from which we came

File: admin_app_2/src/utils/auth.tsx
Comment (amiraliemami) [reply admin_app_2/src/utils/auth.tsx:30]: Hmm interesting - sounds good

File: admin_app_2/src/utils/auth.tsx
Comment (amiraliemami) [reply admin_app_2/src/utils/auth.tsx:35]: I think the second option sounds like a good compromise right now? It'll behave as the user expects and can be made secure later. Maybe we can make that change with a big warning comment above the code reminding us to change it ASAP (tickets sometimes get lost in the backlog and this seems important).

**Review by amiraliemami — APPROVED** [review]
Thanks for the responses. Approving to unblock merging for you tonight but two things to fix before merge:
1. Accept suggestions on button links
2. Auth state loss on refresh

Thanks!

File: admin_app_2/src/app/content/page.tsx
Comment (sidravi1) [reply admin_app_2/src/app/content/page.tsx:57]: Ah got it. Thanks!

File: admin_app_2/src/utils/auth.tsx
Comment (sidravi1) [reply admin_app_2/src/utils/auth.tsx:53]: Fixed!

File: admin_app_2/src/utils/auth.tsx
Comment (sidravi1) [reply admin_app_2/src/utils/auth.tsx:35]: ok! Fixed

File: admin_app_2/src/app/content/layout.tsx
Comment (sidravi1) [reply admin_app_2/src/app/content/layout.tsx:3]: Thanks! Fixed

**Pull request merged**
